pkg/cluster/update_test.go

```diff
@@ -83,3 +98,426 @@ func TestFilterOldVMs(t *testing.T) {
+						VirtualMachineScaleSetVMProperties: &compute.VirtualMachineScaleSetVMProperties{
+							OsProfile: &compute.OSProfile{
+								ComputerName: to.StringPtr("master-000000"),
+							},
+						},
+					},
+				},
+				"infra": {
+					{
+						Name: to.StringPtr("ss-infra_0"),
+						VirtualMachineScaleSetVMProperties: &compute.VirtualMachineScaleSetVMProperties{
+							OsProfile: &compute.OSProfile{
+								ComputerName: to.StringPtr("infra-000000"),
+							},
+						},
+					},
+				},
+				"compute": {
+					{
+						Name: to.StringPtr("ss-compute_0"),
+						VirtualMachineScaleSetVMProperties: &compute.VirtualMachineScaleSetVMProperties{
+							OsProfile: &compute.OSProfile{
+								ComputerName: to.StringPtr("compute-000000"),
+							},
+						},
+					},
+					{
+						Name: to.StringPtr("ss-compute_1"),
+						VirtualMachineScaleSetVMProperties: &compute.VirtualMachineScaleSetVMProperties{
+							OsProfile: &compute.OSProfile{
+								ComputerName: to.StringPtr("compute-000001"),
+							},
+						},
+						InstanceID: to.StringPtr("0123456"),
+					},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+			gmc := gomock.NewController(t)
+			defer gmc.Finish()
+			kubeclient := mock_kubeclient.NewMockKubeclient(gmc)
+			virtualMachineScaleSetsClient := mock_azureclient.NewMockVirtualMachineScaleSetsClient(gmc)
+			virtualMachineScaleSetVMsClient := mock_azureclient.NewMockVirtualMachineScaleSetVMsClient(gmc)
+			mockListVMs(ctx, gmc, virtualMachineScaleSetVMsClient, tt.cs, "master", testRg, tt.vmsBefore["master"], nil)
+			mockListVMs(ctx, gmc, virtualMachineScaleSetVMsClient, tt.cs, "infra", testRg, tt.vmsBefore["infra"], nil)
+			mockListVMs(ctx, gmc, virtualMachineScaleSetVMsClient, tt.cs, "compute", testRg, tt.vmsBefore["compute"], nil)
+
+			if tt.expectDrain != nil {
+				for comp, scalesetName := range tt.expectDrain {
+					kubeclient.EXPECT().Drain(ctx, gomock.Any(), comp)
+					arc := autorest.NewClientWithUserAgent("unittest")
+					virtualMachineScaleSetVMsClient.EXPECT().Client().Return(arc)
+					req, _ := http.NewRequest("delete", "http://example.com", nil)
+					fakeResp := http.Response{Request: req, StatusCode: 200}
+					ft, _ := azure.NewFutureFromResponse(&fakeResp)
+					vFt := compute.VirtualMachineScaleSetVMsDeleteFuture{Future: ft}
+					virtualMachineScaleSetVMsClient.EXPECT().Delete(ctx, testRg, scalesetName, gomock.Any()).Return(vFt, nil)
+				}
+			}
+			u := &simpleUpgrader{
+				vmc:        virtualMachineScaleSetVMsClient,
+				ssc:        virtualMachineScaleSetsClient,
+				kubeclient: kubeclient,
+				log:        logrus.NewEntry(logrus.StandardLogger()),
+			}
+			got, err := u.getNodesAndDrain(ctx, tt.cs)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("simpleUpgrader.getNodesAndDrain() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
```

Comment:
Compare the actual error messages instead of expecting a random error

Reply:
i don't have error testing here just yet, just the different logical paths..